File: C++/13-Heap(PriorityQueue)/Problems/3-Median-of-TwoSortedArrays/Median-of-TwoSortedArraysPro.js
/*
Problem: Median of Two Sorted Arrays

Given two sorted arrays nums1 and nums2 of size m and n respectively, return the median of the two sorted arrays.

The overall run time complexity should be O(log (m+n)).

Example 1:
Input: nums1 = [1,3], nums2 = [2]
Output: 2.00000
Explanation: merged array = [1,2,3] and median is 2.

Example 2:
Input: nums1 = [1,2], nums2 = [3,4]
Output: 2.50000
Explanation: merged array = [1,2,3,4] and median is (2 + 3) / 2 = 2.5.

Constraints:
nums1.length == m
nums2.length == n
0 <= m <= 1000
0 <= n <= 1000
1 <= m + n <= 2000
-10^6 <= nums1[i], nums2[i] <= 10^6
*/

/**
 * Approach 1: Merge and Find Median
 * - Merge the two sorted arrays into a single sorted array.
 * - Calculate the median based on the length of the merged array.
 * - Time Complexity: O(m+n)
 * - Space Complexity: O(m+n)
 */
const findMedianSortedArrays_merge = (nums1, nums2) => {
    const mergedArray = [];
    let i = 0;
    let j = 0;

    while (i < nums1.length && j < nums2.length) {
        if (nums1[i] <= nums2[j]) {
            mergedArray.push(nums1[i]);
            i++;
        } else {
            mergedArray.push(nums2[j]);
            j++;
        }
    }

    while (i < nums1.length) {
        mergedArray.push(nums1[i]);
        i++;
    }

    while (j < nums2.length) {
        mergedArray.push(nums2[j]);
        j++;
    }

    const n = mergedArray.length;
    if (n % 2 === 0) {
        return (mergedArray[n / 2 - 1] + mergedArray[n / 2]) / 2;
    } else {
        return mergedArray[Math.floor(n / 2)];
    }
};

/**
 * Approach 2: Merge and Find Median (Optimized)
 * - Merge the two sorted arrays, but only store the elements needed to calculate the median.
 * - This avoids storing the entire merged array, reducing space complexity.
 * - Time Complexity: O(m+n)
 * - Space Complexity: O(1) - Optimized version uses constant space.
 */
const findMedianSortedArrays_mergeOptimized = (nums1, nums2) => {
    const m = nums1.length;
    const n = nums2.length;
    const total = m + n;
    const half = Math.floor(total / 2);
    let i = 0;
    let j = 0;
    let mid1 = 0;
    let mid2 = 0;

    for (let count = 0; count <= half; count++) {
        mid1 = mid2; // Store the previous mid2
        if (i < m && (j >= n || nums1[i] <= nums2[j])) {
            mid2 = nums1[i];
            i++;
        } else {
            mid2 = nums2[j];
            j++;
        }
    }

    if (total % 2 === 0) {
        return (mid1 + mid2) / 2.0;
    } else {
        return mid2;
    }
};


/**
 * Approach 3: Binary Search
 * - Use binary search on the smaller array to find the partition point.
 * - This approach is more efficient, especially when the arrays have significantly different sizes.
 * - Time Complexity: O(log(min(m, n)))
 * - Space Complexity: O(1)
 */
const findMedianSortedArrays_binarySearch = (nums1, nums2) => {
    const m = nums1.length;
    const n = nums2.length;
    if (m > n) {
        return findMedianSortedArrays_binarySearch(nums2, nums1); // Ensure nums1 is the smaller array
    }

    let low = 0;
    let high = m;
    const half = Math.floor((m + n + 1) / 2);

    while (low <= high) {
        const partitionX = Math.floor((low + high) / 2);
        const partitionY = half - partitionX;

        const maxLeftX = partitionX === 0 ? -Infinity : nums1[partitionX - 1];
        const minRightX = partitionX === m ? Infinity : nums1[partitionX];

        const maxLeftY = partitionY === 0 ? -Infinity : nums2[partitionY - 1];
        const minRightY = partitionY === n ? Infinity : nums2[partitionY];

        if (maxLeftX <= minRightY && maxLeftY <= minRightX) {
            if ((m + n) % 2 === 0) {
                return (Math.max(maxLeftX, maxLeftY) + Math.min(minRightX, minRightY)) / 2.0;
            } else {
                return Math.max(maxLeftX, maxLeftY);
            }
        } else if (maxLeftX > minRightY) {
            high = partitionX - 1;
        } else {
            low = partitionX + 1;
        }
    }
    return 0; //Should not reach here, but added for the sake of completeness.
};

/**
 * Approach 4: Using a Min Heap (Priority Queue)
 * - Insert all elements from both arrays into a min heap.
 * - Pop elements from the heap until we reach the median.
 * - Time Complexity: O((m+n)log(m+n))  due to heap insertion
 * - Space Complexity: O(m+n)
 */

class MinHeap {
    constructor() {
        this.heap = [];
    }

    insert(value) {
        this.heap.push(value);
        this.bubbleUp(this.heap.length - 1);
    }

    bubbleUp(index) {
        while (index > 0) {
            const parentIndex = Math.floor((index - 1) / 2);
            if (this.heap[index] < this.heap[parentIndex]) {
                [this.heap[index], this.heap[parentIndex]] = [this.heap[parentIndex], this.heap[index]];
                index = parentIndex;
            } else {
                break;
            }
        }
    }

    extractMin() {
        if (this.heap.length === 0) return null;
        if (this.heap.length === 1) return this.heap.pop();

        const min = this.heap[0];
        this.heap[0] = this.heap.pop();
        this.bubbleDown(0);
        return min;
    }

    bubbleDown(index) {
        while (true) {
            const leftChild = 2 * index + 1;
            const rightChild = 2 * index + 2;
            let smallest = index;

            if (leftChild < this.heap.length && this.heap[leftChild] < this.heap[smallest]) {
                smallest = leftChild;
            }

            if (rightChild < this.heap.length && this.heap[rightChild] < this.heap[smallest]) {
                smallest = rightChild;
            }

            if (smallest !== index) {
                [this.heap[index], this.heap[smallest]] = [this.heap[smallest], this.heap[index]];
                index = smallest;
            } else {
                break;
            }
        }
    }

    size() {
        return this.heap.length;
    }
}

const findMedianSortedArrays_heap = (nums1, nums2) => {
    const minHeap = new MinHeap();
    for (const num of nums1) {
        minHeap.insert(num);
    }
    for (const num of nums2) {
        minHeap.insert(num);
    }

    const total = nums1.length + nums2.length;
    const half = Math.floor(total / 2);

    for (let i = 0; i < half; i++) {
        minHeap.extractMin();
    }

    if (total % 2 === 0) {
        return (minHeap.extractMin() + minHeap.extractMin()) / 2;
    } else {
        return minHeap.extractMin();
    }
};

/**
 * Approach 5: Using Javascript Array Sort (Not Recommended for Performance)
 * - Concatenate the two arrays and sort the result.
 * - Find the median of the sorted array.
 * - This is the simplest to implement but has the worst performance.
 * - Time Complexity: O((m+n)log(m+n)) due to sorting
 * - Space Complexity: O(m+n)
 */
const findMedianSortedArrays_sort = (nums1, nums2) => {
    const mergedArray = nums1.concat(nums2);
    mergedArray.sort((a, b) => a - b); // Sort numerically
    const n = mergedArray.length;
    if (n % 2 === 0) {
        return (mergedArray[n / 2 - 1] + mergedArray[n / 2]) / 2;
    } else {
        return mergedArray[Math.floor(n / 2)];
    }
};



// Example Usage and Console Logs
const nums1_1 = [1, 3];
const nums2_1 = [2];
console.log("Example 1:");
console.log("Approach 1 (Merge):", findMedianSortedArrays_merge(nums1_1, nums2_1));
console.log("Approach 2 (Merge Optimized):", findMedianSortedArrays_mergeOptimized(nums1_1, nums2_1));
console.log("Approach 3 (Binary Search):", findMedianSortedArrays_binarySearch(nums1_1, nums2_1));
console.log("Approach 4 (Heap):", findMedianSortedArrays_heap(nums1_1, nums2_1));
console.log("Approach 5 (Sort):", findMedianSortedArrays_sort(nums1_1, nums2_1));

const nums1_2 = [1, 2];
const nums2_2 = [3, 4];
console.log("\nExample 2:");
console.log("Approach 1 (Merge):", findMedianSortedArrays_merge(nums1_2, nums2_2));
console.log("Approach 2 (Merge Optimized):", findMedianSortedArrays_mergeOptimized(nums1_2, nums2_2));
console.log("Approach 3 (Binary Search):", findMedianSortedArrays_binarySearch(nums1_2, nums2_2));
console.log("Approach 4 (Heap):", findMedianSortedArrays_heap(nums1_2, nums2_2));
console.log("Approach 5 (Sort):", findMedianSortedArrays_sort(nums1_2, nums2_2));

const nums1_3 = [0, 0];
const nums2_3 = [0, 0];
console.log("\nExample 3:");
console.log("Approach 1 (Merge):", findMedianSortedArrays_merge(nums1_3, nums2_3));
console.log("Approach 2 (Merge Optimized):", findMedianSortedArrays_mergeOptimized(nums1_3, nums2_3));
console.log("Approach 3 (Binary Search):", findMedianSortedArrays_binarySearch(nums1_3, nums2_3));
console.log("Approach 4 (Heap):", findMedianSortedArrays_heap(nums1_3, nums2_3));
console.log("Approach 5 (Sort):", findMedianSortedArrays_sort(nums1_3, nums2_3));

const nums1_4 = [2];
const nums2_4 = [];
console.log("\nExample 4:");
console.log("Approach 1 (Merge):", findMedianSortedArrays_merge(nums1_4, nums2_4));
console.log("Approach 2 (Merge Optimized):", findMedianSortedArrays_mergeOptimized(nums1_4, nums2_4));
console.log("Approach 3 (Binary Search):", findMedianSortedArrays_binarySearch(nums1_4, nums2_4));
console.log("Approach 4 (Heap):", findMedianSortedArrays_heap(nums1_4, nums2_4));
console.log("Approach 5 (Sort):", findMedianSortedArrays_sort(nums1_4, nums2_4));

File: C++/13-Heap(PriorityQueue)/Problems/3-Median-of-TwoSortedArrays/Median-of-TwoSortedArraysPro.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <queue>
#include <limits>
#include <iomanip> // Required for std::setprecision

using namespace std;

// Function to print the array (for debugging purposes)
void printArray(const vector<int>& arr, const string& name) {
    cout << name << ": ";
    for (int val : arr) {
        cout << val << " ";
    }
    cout << endl;
}

// 1. Brute Force Approach: Merge and Sort
//   - Time Complexity: O(m + n) for merging, O((m + n) log (m + n)) for sorting
//   - Space Complexity: O(m + n) for the merged array
//   - Explanation: This is the most straightforward approach.  Merge the two sorted arrays into a single sorted array, and then find the median.  While simple, it's not the most efficient.
double findMedianSortedArrays_BruteForce(const vector<int>& nums1, const vector<int>& nums2) {
    vector<int> mergedArray;
    mergedArray.reserve(nums1.size() + nums2.size()); // Pre-allocate memory for efficiency

    // Merge the two sorted arrays
    int i = 0, j = 0;
    while (i < nums1.size() && j < nums2.size()) {
        if (nums1[i] <= nums2[j]) {
            mergedArray.push_back(nums1[i++]);
        } else {
            mergedArray.push_back(nums2[j++]);
        }
    }
    while (i < nums1.size()) {
        mergedArray.push_back(nums1[i++]);
    }
    while (j < nums2.size()) {
        mergedArray.push_back(nums2[j++]);
    }

    // Sort the merged array (not needed because merge process already sorts)
    // sort(mergedArray.begin(), mergedArray.end());  // Redundant, already sorted

    int mergedSize = mergedArray.size();
    if (mergedSize % 2 == 0) {
        // Even number of elements: median is the average of the two middle elements
        return (double)(mergedArray[mergedSize / 2 - 1] + mergedArray[mergedSize / 2]) / 2.0;
    } else {
        // Odd number of elements: median is the middle element
        return (double)mergedArray[mergedSize / 2];
    }
}

// 2. Merge and Iterate Approach
//   - Time Complexity: O(m + n)
//   - Space Complexity: O(1) -  Uses a constant amount of extra space.
//   - Explanation:  Similar to the brute force approach, but instead of storing the entire merged array, we only keep track of the two middle elements (or one middle element if the total count is odd) as we iterate through the arrays. This significantly reduces the space complexity.
double findMedianSortedArrays_MergeIterate(const vector<int>& nums1, const vector<int>& nums2) {
    int m = nums1.size();
    int n = nums2.size();
    int total = m + n;
    int mid = total / 2;
    int i = 0, j = 0, count = 0;
    int first = 0, second = 0; // Store the two middle elements

    while (count <= mid) {
        second = first; // Keep track of the previous "first"
        if (i < m && j < n) {
            if (nums1[i] <= nums2[j]) {
                first = nums1[i++];
            } else {
                first = nums2[j++];
            }
        } else if (i < m) {
            first = nums1[i++];
        } else {
            first = nums2[j++];
        }
        count++;
    }

    if (total % 2 == 0) {
        return (double)(first + second) / 2.0;
    } else {
        return (double)first;
    }
}

// 3. Binary Search Approach
//   - Time Complexity: O(log(min(m, n)))
//   - Space Complexity: O(1)
//   - Explanation: This is the most efficient approach. It uses binary search to find the correct partition of the two arrays such that the median can be calculated from the elements around the partition.  It works by ensuring that the number of elements on the left side of the partition is equal to the number of elements on the right side.
double findMedianSortedArrays_BinarySearch(const vector<int>& nums1, const vector<int>& nums2) {
    int m = nums1.size();
    int n = nums2.size();
    // Ensure nums1 is the smaller array to optimize binary search
    if (m > n) {
        return findMedianSortedArrays_BinarySearch(nums2, nums1);
    }

    int low = 0, high = m;
    while (low <= high) {
        int partitionX = (low + high) / 2;
        int partitionY = ((m + n + 1) / 2) - partitionX;

        // Get max/min elements for left and right sides.  Handle edge cases.
        int maxLeftX = (partitionX == 0) ? numeric_limits<int>::min() : nums1[partitionX - 1];
        int minRightX = (partitionX == m) ? numeric_limits<int>::max() : nums1[partitionX];
        int maxLeftY = (partitionY == 0) ? numeric_limits<int>::min() : nums2[partitionY - 1];
        int minRightY = (partitionY == n) ? numeric_limits<int>::max() : nums2[partitionY];

        // Check if we have found the correct partition
        if (maxLeftX <= minRightY && maxLeftY <= minRightX) {
            // Calculate the median
            if ((m + n) % 2 == 0) {
                return (double)(max(maxLeftX, maxLeftY) + min(minRightX, minRightY)) / 2.0;
            } else {
                return (double)max(maxLeftX, maxLeftY);
            }
        } else if (maxLeftX > minRightY) {
            // Move towards the left side of nums1
            high = partitionX - 1;
        } else {
            // Move towards the right side of nums1
            low = partitionX + 1;
        }
    }
    // If the input arrays are not sorted, the loop will exit and return 0.0
    return 0.0;
}

// 4. Using Priority Queues (Max-Heap and Min-Heap)
//   - Time Complexity: O((m+n)log(m+n)).  Adding elements to heaps is log(n).
//   - Space Complexity: O(m + n) in the worst case, to store all elements in the heaps.
//   - Explanation: This approach uses two priority queues (heaps): a max-heap to store the left half of the merged array and a min-heap to store the right half.  This method maintains the property that all elements in the max-heap are less than or equal to all elements in the min-heap.
double findMedianSortedArrays_Heaps(const vector<int>& nums1, const vector<int>& nums2) {
    priority_queue<int> maxHeap; // Max-heap for the left part
    priority_queue<int, vector<int>, greater<int>> minHeap; // Min-heap for the right part

    // Insert elements from both arrays into the heaps
    for (int num : nums1) {
        maxHeap.push(num);
    }
    for (int num : nums2) {
        maxHeap.push(num);
    }

     // Rebalance the heaps
    while (maxHeap.size() > minHeap.size() + 1) {
        minHeap.push(maxHeap.top());
        maxHeap.pop();
    }
    while (minHeap.size() > maxHeap.size()) {
        maxHeap.push(minHeap.top());
        minHeap.pop();
    }

    // Calculate the median
    if ((maxHeap.size() + minHeap.size()) % 2 == 0) {
        return (double)(maxHeap.top() + minHeap.top()) / 2.0;
    } else {
        return (double)maxHeap.top();
    }
}

// 5. Recursive Approach (Divide and Conquer)
//   - Time Complexity: O(log(m + n)) on average, O(m + n) in the worst case
//   - Space Complexity: O(log(m + n)) for the recursion stack
//   - Explanation: This approach recursively divides the arrays to find the kth smallest element, where k is (m+n)/2.  It's based on the idea that the median is the kth smallest element.  This method can be tricky to implement correctly due to the recursive logic.
double findMedianSortedArrays_Recursive(const vector<int>& nums1, int start1, int end1,
                                            const vector<int>& nums2, int start2, int end2, int k) {
    // Handle base cases for recursion
    if (start1 > end1) {
        return (double)nums2[start2 + k - 1];
    }
    if (start2 > end2) {
        return (double)nums1[start1 + k - 1];
    }
    if (k == 1) {
        return min(nums1[start1], nums2[start2]);
    }

    int mid1 = (start1 + end1) / 2;
    int mid2 = (start2 + end2) / 2;

    // Handle cases where mid1 or mid2 goes out of bounds
    int len1 = end1 - start1 + 1;
    int len2 = end2 - start2 + 1;
    int val1 = (mid1 < start1 + len1) ? nums1[mid1] : numeric_limits<int>::max();
    int val2 = (mid2 < start2 + len2) ? nums2[mid2] : numeric_limits<int>::max();

    if (val1 <= val2) {
        if (k > (mid1 - start1) + (mid2 - start2) + 1) {
            return findMedianSortedArrays_Recursive(nums1, mid1 + 1, end1, nums2, start2, end2,
                                                    k - (mid1 - start1) - 1);
        } else {
            return findMedianSortedArrays_Recursive(nums1, start1, end1, nums2, start2, mid2 - 1, k);
        }
    } else {
        if (k > (mid1 - start1) + (mid2 - start2) + 1) {
            return findMedianSortedArrays_Recursive(nums1, start1, end1, nums2, mid2 + 1, end2,
                                                    k - (mid2 - start2) - 1);
        } else {
            return findMedianSortedArrays_Recursive(nums1, start1, mid1 - 1, nums2, start2, end2, k);
        }
    }
}

double findMedianSortedArrays_Recursive(const vector<int>& nums1, const vector<int>& nums2) {
    int m = nums1.size();
    int n = nums2.size();
    int total = m + n;
    if (total % 2 == 0) {
        return (findMedianSortedArrays_Recursive(nums1, 0, m - 1, nums2, 0, n - 1, total / 2) +
                findMedianSortedArrays_Recursive(nums1, 0, m - 1, nums2, 0, n - 1, total / 2 + 1)) / 2.0;
    } else {
        return findMedianSortedArrays_Recursive(nums1, 0, m - 1, nums2, 0, n - 1, total / 2 + 1);
    }
}

int main() {
    // Example usage
    vector<int> nums1 = {1, 3, 8, 9, 15};
    vector<int> nums2 = {7, 11, 18, 19, 21, 25};

    cout << fixed << setprecision(2); //  Control decimal precision for output

    cout << "Brute Force: " << findMedianSortedArrays_BruteForce(nums1, nums2) << endl;
    cout << "Merge and Iterate: " << findMedianSortedArrays_MergeIterate(nums1, nums2) << endl;
    cout << "Binary Search: " << findMedianSortedArrays_BinarySearch(nums1, nums2) << endl;
    cout << "Heaps: " << findMedianSortedArrays_Heaps(nums1, nums2) << endl;
    cout << "Recursive: " << findMedianSortedArrays_Recursive(nums1, nums2) << endl;

    return 0;
}
